From: Michal Kubecek <mkubecek@suse.cz>
Subject: novfs: fix off-by-one allocation error
References: bnc#669378
Patch-mainline: no

Allocation of buffer in novfs_listx_file_info() missed one byte for
null byte mark which caused buffer overflow with path of exactly
108 bytes, leading to rewriting first byte of next slab object and
corrupting DirCache linked lists.

Reported-by: Jeremy Meldrum <jmeldrum@novell.com>
Signed-off-by: Michal Kubecek <mkubecek@suse.cz>
Acked-by: Jan Kara <jack@suse.cz>
---
 fs/novfs/file.c |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/fs/novfs/file.c b/fs/novfs/file.c
index 0bd60a6..e8dfc69 100644
--- a/fs/novfs/file.c
+++ b/fs/novfs/file.c
@@ -414,7 +414,7 @@ int novfs_listx_file_info(char *Path, char *buffer, ssize_t buffer_size, ssize_t
 
 	*dataLen = 0;
 	cmdlen = offsetof(struct novfs_verify_file_request, path) + pathlen;
-	cmd = (struct novfs_verify_file_request *)kmalloc(cmdlen, GFP_KERNEL);
+	cmd = (struct novfs_verify_file_request *) kmalloc(cmdlen + 1, GFP_KERNEL);
 	if (cmd) {
 		cmd->Command.CommandType = VFS_COMMAND_LIST_EXTENDED_ATTRIBUTES;
 		cmd->Command.SequenceNumber = 0;
-- 
1.7.3.4

